### ISSUE TITLE
Transaction is not rolled back on error during migration

### ISSUE THREAD

**Issue by hannesvdvreken** [typeorm/typeorm#9187]
### Expected Behavior

When running a migration file with transaction "each" option, the transaction should be rolled back on error.

### Actual Behavior

When running a migration file with transaction "each" option, the transaction is not rolled back on error.

```
query: START TRANSACTION
query: ALTER TABLE `users` DROP COLUMN `email`
query: ALTER TABLE `users` DROP COLUMN `non-existent-column`
error: Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version...
Migration "RemoveUserData1643809966646" failed
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |  node-docker  |
| Node.js version     | 16.15.0  | 
| Typescript version  | 4.4.3 | 
| TypeORM version     | 0.3.6  | 

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | yes            |
| `aurora-postgres` | yes            |
| `better-sqlite3`  | yes            |
| `cockroachdb`     | yes            |
| `cordova`         | yes            |
| `expo`            | yes            |
| `mongodb`         | no            |
| `mysql`           | yes            |
| `nativescript`    | no          |
| `oracle`          | yes           |
| `postgres`        | yes           |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | yes            |
| `sqlite-abstract` | yes            |
| `sqljs`           | yes           |
| `sqlserver`       | yes            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✅ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by abdallahbedir2** [comment]
@ertl When a transaction migration fails, why not running the corresponding `down()` queries for the executed queries before an error occur

i.e. the following was run successfully but we need to revert it as all queries not succeed
```
ALTER TABLE `users` DROP COLUMN `email`
```  

then we have to run the corresponding `down()` query which might be 
```
ALTER TABLE `users` ADD COLUMN `email`
```

What do you think?
I had as similar issue [here](https://stackoverflow.com/questions/77842760/typeorm-rollback-cannot-execute-down-fn-when-migration-run-fails)

**Comment by ertl** [comment]
@abdallahbedir2 Yes that could work. We would then have to find and execute the appropriate down statements in the correct order. (possibly in a separate transaction?)
What concerns is, how we want to deal with DML statements. All executed DML statements should be reliably rolled back. This means that no down statement needs to be executed for all DML statements.
For this we probably need a function to differentiate between DML and DDL statements.
My first approach would be to check whether the statement begins with SELECT, INSERT, UPDATE, DELETE, MERGE in order to determine whether it is a DML statement. But I'm not sure whether all DML statements really fall under this or whether we have forgotten some.

**Comment by chardskarth** [comment]
Sadly, this happens to me using `type: mysql`. But `type: postgres` rollsback fine.

**Comment by chardskarth** [comment]
Is there a workaround? Adding "--transaction each" doesnt seem to work for me too when using 'mysql'

**Comment by Eldow** [comment]
I can reproduce the issue with the mysql driver.

**Comment by chardskarth** [comment]
Unfortunately, I think its simply because of how MariaDB/MySQL works. 

"DDL" statements (iiuc, are statements that change structure of tables and columns), are immediately taking effect regardless of the transaction. 

It's a bummer. I think thats just how db engine of mysql work. I'm opting to use postgres instead.

**Comment by Motii1** [comment]
> Unfortunately, I think its simply because of how MariaDB/MySQL works.
> 
> "DDL" statements (iiuc, are statements that change structure of tables and columns), are immediately taking effect regardless of the transaction.
> 
> It's a bummer. I think thats just how db engine of mysql work. I'm opting to use postgres instead.

I am using MySQL and unfortunately transaction with not only DDL statements but also `UPDATE` statements was not rolled back on my production server

**Comment by ertl** [comment]
I believe this remains a major issue. When a database migration script fails , it leaves the system in an inconsistent state that requires manual cleanup. This is especially critical in production environments. For this reason, I think the migration logic should be reconsidered and reworked.

Unfortunately, I don’t have the rights to reopen the issue myself.

**Comment by gioboa** [comment]
Thanks for the heads up @ertl

**Comment by thecoolestcoderrushil** [comment]
a failed migration that leaves half-applied changes with no rollback is one of the scariest failure modes, because you can no longer trust what state the database is in. rehearsing migrations against a production-shaped twin first shows which statements will fail and what state they leave behind. antifailure.dev (https://antifailure.dev) builds disposable postgres twins for this kind of migration rehearsal and is open source and free to try, with demos available for teams that want one.
